### DISCUSSION TITLE
Post-processing textual content with typeset js

### DISCUSSION THREAD

**Discussion by davenquinn** [mdx-js/mdx#1487] [Q&A] [answered]
I'm interested in using [Typeset js](https://typeset.lllllllllllllllll.com/) as part of a pipeline for MDX, to improve some visual elements in markdown running text.

I've read through a few examples of how to create plugins (e.g. unified JS plugins) for different parts of the MDX system, but I'm still a little puzzled what the best way to proceed here would be, since there are several ways to hook into the rendering. 
Typeset seems to work best with a static input of html; this is of course impractical in a react/JSX context. Should I be thinking of hooking in somewhere at a single-component level, or is there a place I can modify all textual dom nodes while keeping nested components intact if possible?

**Comment by ChristianMurphy** [comment]
> Typeset seems to work best with a static input of html; this is of course impractical in a react/JSX context

you're right, Typeset wants to work with static HTML.
At no point in the mdx pipeline is the content static HTML.
If you want to use Typeset as is, you'd need to render the MDX content to react then render the react to static HTML (https://reactjs.org/docs/react-dom-server.html#rendertostaticmarkup), then apply Typeset js after.

**Comment by ChristianMurphy** [comment, marked as answer]
> Should I be thinking of hooking in somewhere at a single-component level, or is there a place I can modify all textual dom nodes while keeping nested components intact if possible?

It really depends on what you want to do.
The closest to textual dom nodes MDX gets is the hast+mdx phase https://mdxjs.com/advanced/ast
which can be accessed with rehype plugins https://mdxjs.com/advanced/plugins
Unified/MDX provides a guide on creating plugins here: https://unifiedjs.com/learn/
